### TASK 1
#### IMPORTING & CLEANING FINANCIAL DATA
#### Companies - Tesla, Inc., Apple Inc., Microsoft Corporation.
##### Import data
##### Calculate yearly trend for each metric

In [1]:
# import libraries
import pandas as pd
import numpy as np

In [2]:
# Import csv data
BCG_df = pd.read_csv('BCG_Financials.csv')
BCG_df.head()

Company  Year  Total Revenue ($M)  Net Income ($M)  Total Assets ($M)  \
0      Apple  2022              394328            99803             352755   
1      Apple  2023              383285            96995             352583   
2      Apple  2024              391035            93736             364980   
3  Microsoft  2022              198270            72738             364840   
4  Microsoft  2023              211915            72361             411976   

   Total Liabilities ($M)  Cash Flow from Operating Activities ($M)  
0                  302083                                    122151  
1                  290437                                    110543  
2                  308030                                    118254  
3                  198298                                     89035  
4                  205753                                     87582

In [3]:
BCG_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   Company                                   9 non-null      object
 1   Year                                      9 non-null      int64 
 2   Total Revenue ($M)                        9 non-null      int64 
 3   Net Income ($M)                           9 non-null      int64 
 4   Total Assets ($M)                         9 non-null      int64 
 5   Total Liabilities ($M)                    9 non-null      int64 
 6   Cash Flow from Operating Activities ($M)  9 non-null      int64 
dtypes: int64(6), object(1)
memory usage: 636.0+ bytes


In [4]:
BCG_df.describe()

Year  Total Revenue ($M)  Net Income ($M)  Total Assets ($M)  \
count     9.000000            9.000000         9.000000           9.000000   
mean   2023.000000       233320.000000     62053.666667      296702.555556   
std       0.866025       130006.968331     39095.864110      153240.424606   
min    2022.000000        81462.000000      7153.000000       82338.000000   
25%    2022.000000        97690.000000     14974.000000      122070.000000   
50%    2023.000000       211915.000000     72738.000000      352755.000000   
75%    2024.000000       383285.000000     93736.000000      364980.000000   
max    2024.000000       394328.000000     99803.000000      512163.000000   

       Total Liabilities ($M)  Cash Flow from Operating Activities ($M)  
count                9.000000                                  9.000000  
mean            186236.222222                              76557.333333  
std             114506.220759                              48272.769855  
min              36440.000000                              13256.000000  
25%              48390.000000                              14923.000000  
50%             205753.000000                              89035.000000  
75%             290437.000000                             118254.000000  
max             308030.000000                             122151.000000

#### Total Revenue Growth rate
##### Apple
- Apple experienced a 2.8% reduction in its total revenue.
- However, in 2024 they experienced a 2.02% growth from the previous reduction.
##### Microsoft
- Microsoft has experienced revenue growth wit 6.8% in 2023 and 15% in 2024
##### Tesla
- In 2023, Tesla experienced 18.8% revenue growth.
- However, in 2024, the 0.9% revenue growth was small compared to the previous year's.

In [23]:
# Calculate yearly trend for all metrics and add as a new column
BCG_df['Revenue_growth (%)'] = (BCG_df.groupby("Company")["Total Revenue ($M)"].pct_change() * 100).round(3)

BCG_df

Company  Year  Total Revenue ($M)  Net Income ($M)  Total Assets ($M)  \
0      Apple  2022              394328            99803             352755   
1      Apple  2023              383285            96995             352583   
2      Apple  2024              391035            93736             364980   
3  Microsoft  2022              198270            72738             364840   
4  Microsoft  2023              211915            72361             411976   
5  Microsoft  2024              245122            88136             512163   
6      Tesla  2022               81462            12587              82338   
7      Tesla  2023               96773            14974             106618   
8      Tesla  2024               97690             7153             122070   

   Total Liabilities ($M)  Cash Flow from Operating Activities ($M)  \
0                  302083                                    122151   
1                  290437                                    110543   
2                  308030                                    118254   
3                  198298                                     89035   
4                  205753                                     87582   
5                  243686                                    118548   
6                   36440                                     14724   
7                   43009                                     13256   
8                   48390                                     14923   

   Revenue_growth (%)  NetIncome_growth (%)  Assets_growth (%)  \
0                 NaN              0.000000           0.000000   
1              -2.800             -2.813543          -0.048759   
2               2.022             -3.359967           3.516052   
3                 NaN              0.000000           0.000000   
4               6.882             -0.518299          12.919636   
5              15.670             21.800417          24.318650   
6                 NaN              0.000000           0.000000   
7              18.795             18.964010          29.488207   
8               0.948            -52.230533          14.492862   

   Liabilities_growth (%)  Operations cashflow growth (%)  Net profit margin %  
0                0.000000                        0.000000            25.309641  
1               -3.855232                       -9.502992            25.306234  
2                6.057424                        6.975566            23.971256  
3                0.000000                        0.000000            36.686337  
4                3.759493                       -1.631942            34.146238  
5               18.436183                       35.356580            35.955973  
6                0.000000                        0.000000            15.451376  
7               18.026894                       -9.970117            15.473324  
8               12.511335                       12.575438             7.322141

#### Net Income Growth rate
##### Apple
- Apple experienced a 2.8% reduction in its net income in 2023.
- And a consecutive 3.3% reduction in 2024.
##### Microsoft
- Microsoft has experienced a 0.5% net income reduction in 2023.
- And a 21.8% net income increase in 2024 (The highest growth).
##### Tesla
- In 2023, Tesla experienced 18.9% (The highest growth)net income growth.
- However, in 2024 there was a huge reduction of its net income, 52.2% (The highest reduction overall).

In [24]:
BCG_df['NetIncome_growth (%)'] = (BCG_df.groupby("Company")["Net Income ($M)"].pct_change() * 100).round(3)

BCG_df

Company  Year  Total Revenue ($M)  Net Income ($M)  Total Assets ($M)  \
0      Apple  2022              394328            99803             352755   
1      Apple  2023              383285            96995             352583   
2      Apple  2024              391035            93736             364980   
3  Microsoft  2022              198270            72738             364840   
4  Microsoft  2023              211915            72361             411976   
5  Microsoft  2024              245122            88136             512163   
6      Tesla  2022               81462            12587              82338   
7      Tesla  2023               96773            14974             106618   
8      Tesla  2024               97690             7153             122070   

   Total Liabilities ($M)  Cash Flow from Operating Activities ($M)  \
0                  302083                                    122151   
1                  290437                                    110543   
2                  308030                                    118254   
3                  198298                                     89035   
4                  205753                                     87582   
5                  243686                                    118548   
6                   36440                                     14724   
7                   43009                                     13256   
8                   48390                                     14923   

   Revenue_growth (%)  NetIncome_growth (%)  Assets_growth (%)  \
0                 NaN                   NaN           0.000000   
1              -2.800                -2.814          -0.048759   
2               2.022                -3.360           3.516052   
3                 NaN                   NaN           0.000000   
4               6.882                -0.518          12.919636   
5              15.670                21.800          24.318650   
6                 NaN                   NaN           0.000000   
7              18.795                18.964          29.488207   
8               0.948               -52.231          14.492862   

   Liabilities_growth (%)  Operations cashflow growth (%)  Net profit margin %  
0                0.000000                        0.000000            25.309641  
1               -3.855232                       -9.502992            25.306234  
2                6.057424                        6.975566            23.971256  
3                0.000000                        0.000000            36.686337  
4                3.759493                       -1.631942            34.146238  
5               18.436183                       35.356580            35.955973  
6                0.000000                        0.000000            15.451376  
7               18.026894                       -9.970117            15.473324  
8               12.511335                       12.575438             7.322141

#### Total Assets Growth rate
##### Apple
- Apple experienced a 0.04%% reduction in its total assets.
- However, in 2024 they experienced a 3.5% growth of assets.
##### Microsoft
- Microsoft has experienced asset growth with 12.9% in 2023 and 24.3% in 2024 (The highest).
##### Tesla
- Tesla has experienced asset growth with 29.5% (The highest)in 2023 and 14.5% in 2024.

In [25]:
BCG_df['Assets_growth (%)'] = (BCG_df.groupby("Company")["Total Assets ($M)"].pct_change() * 100).round(3)

BCG_df

Company  Year  Total Revenue ($M)  Net Income ($M)  Total Assets ($M)  \
0      Apple  2022              394328            99803             352755   
1      Apple  2023              383285            96995             352583   
2      Apple  2024              391035            93736             364980   
3  Microsoft  2022              198270            72738             364840   
4  Microsoft  2023              211915            72361             411976   
5  Microsoft  2024              245122            88136             512163   
6      Tesla  2022               81462            12587              82338   
7      Tesla  2023               96773            14974             106618   
8      Tesla  2024               97690             7153             122070   

   Total Liabilities ($M)  Cash Flow from Operating Activities ($M)  \
0                  302083                                    122151   
1                  290437                                    110543   
2                  308030                                    118254   
3                  198298                                     89035   
4                  205753                                     87582   
5                  243686                                    118548   
6                   36440                                     14724   
7                   43009                                     13256   
8                   48390                                     14923   

   Revenue_growth (%)  NetIncome_growth (%)  Assets_growth (%)  \
0                 NaN                   NaN                NaN   
1              -2.800                -2.814             -0.049   
2               2.022                -3.360              3.516   
3                 NaN                   NaN                NaN   
4               6.882                -0.518             12.920   
5              15.670                21.800             24.319   
6                 NaN                   NaN                NaN   
7              18.795                18.964             29.488   
8               0.948               -52.231             14.493   

   Liabilities_growth (%)  Operations cashflow growth (%)  Net profit margin %  
0                0.000000                        0.000000            25.309641  
1               -3.855232                       -9.502992            25.306234  
2                6.057424                        6.975566            23.971256  
3                0.000000                        0.000000            36.686337  
4                3.759493                       -1.631942            34.146238  
5               18.436183                       35.356580            35.955973  
6                0.000000                        0.000000            15.451376  
7               18.026894                       -9.970117            15.473324  
8               12.511335                       12.575438             7.322141

#### Total Liabilities Growth rate
##### Apple
- Apple experienced a 3.8% reduction in its total liabilities.
- However, in 2024 they experienced a 6.05% growth.
##### Microsoft
- Microsoft has experienced revenue growth with 3.7% in 2023 and 18.4% in 2024.
##### Tesla
- Tesla experienced an 18% liabilities increase in 2023 and 12.5% increase in 2024.

In [26]:
BCG_df['Liabilities_growth (%)'] = (BCG_df.groupby("Company")["Total Liabilities ($M)"].pct_change() * 100).round(3)

BCG_df

Company  Year  Total Revenue ($M)  Net Income ($M)  Total Assets ($M)  \
0      Apple  2022              394328            99803             352755   
1      Apple  2023              383285            96995             352583   
2      Apple  2024              391035            93736             364980   
3  Microsoft  2022              198270            72738             364840   
4  Microsoft  2023              211915            72361             411976   
5  Microsoft  2024              245122            88136             512163   
6      Tesla  2022               81462            12587              82338   
7      Tesla  2023               96773            14974             106618   
8      Tesla  2024               97690             7153             122070   

   Total Liabilities ($M)  Cash Flow from Operating Activities ($M)  \
0                  302083                                    122151   
1                  290437                                    110543   
2                  308030                                    118254   
3                  198298                                     89035   
4                  205753                                     87582   
5                  243686                                    118548   
6                   36440                                     14724   
7                   43009                                     13256   
8                   48390                                     14923   

   Revenue_growth (%)  NetIncome_growth (%)  Assets_growth (%)  \
0                 NaN                   NaN                NaN   
1              -2.800                -2.814             -0.049   
2               2.022                -3.360              3.516   
3                 NaN                   NaN                NaN   
4               6.882                -0.518             12.920   
5              15.670                21.800             24.319   
6                 NaN                   NaN                NaN   
7              18.795                18.964             29.488   
8               0.948               -52.231             14.493   

   Liabilities_growth (%)  Operations cashflow growth (%)  Net profit margin %  
0                     NaN                        0.000000            25.309641  
1                  -3.855                       -9.502992            25.306234  
2                   6.057                        6.975566            23.971256  
3                     NaN                        0.000000            36.686337  
4                   3.759                       -1.631942            34.146238  
5                  18.436                       35.356580            35.955973  
6                     NaN                        0.000000            15.451376  
7                  18.027                       -9.970117            15.473324  
8                  12.511                       12.575438             7.322141

#### Total Operations Cash flow Growth rate
##### Apple
- Apple experienced a 9.58% reduction in its total operations cash flow in 2023.
- However, in 2024 they experienced a 6.97% growth from the previous reduction.
##### Microsoft
- Microsoft has experienced a 1.6% reduction in its operations cash flow in 2023.
- While it experienced a 35.4% growth in 2024.
##### Tesla
- In 2023, Tesla experienced 9.9% reduction in operations cash flow.
- However in 2024, it experienced a 12.5% operations cash flow increase.

In [27]:
BCG_df['Operations cashflow growth (%)'] = (BCG_df.groupby("Company")["Cash Flow from Operating Activities ($M)"].pct_change() * 100).round(3)

BCG_df

Company  Year  Total Revenue ($M)  Net Income ($M)  Total Assets ($M)  \
0      Apple  2022              394328            99803             352755   
1      Apple  2023              383285            96995             352583   
2      Apple  2024              391035            93736             364980   
3  Microsoft  2022              198270            72738             364840   
4  Microsoft  2023              211915            72361             411976   
5  Microsoft  2024              245122            88136             512163   
6      Tesla  2022               81462            12587              82338   
7      Tesla  2023               96773            14974             106618   
8      Tesla  2024               97690             7153             122070   

   Total Liabilities ($M)  Cash Flow from Operating Activities ($M)  \
0                  302083                                    122151   
1                  290437                                    110543   
2                  308030                                    118254   
3                  198298                                     89035   
4                  205753                                     87582   
5                  243686                                    118548   
6                   36440                                     14724   
7                   43009                                     13256   
8                   48390                                     14923   

   Revenue_growth (%)  NetIncome_growth (%)  Assets_growth (%)  \
0                 NaN                   NaN                NaN   
1              -2.800                -2.814             -0.049   
2               2.022                -3.360              3.516   
3                 NaN                   NaN                NaN   
4               6.882                -0.518             12.920   
5              15.670                21.800             24.319   
6                 NaN                   NaN                NaN   
7              18.795                18.964             29.488   
8               0.948               -52.231             14.493   

   Liabilities_growth (%)  Operations cashflow growth (%)  Net profit margin %  
0                     NaN                             NaN            25.309641  
1                  -3.855                          -9.503            25.306234  
2                   6.057                           6.976            23.971256  
3                     NaN                             NaN            36.686337  
4                   3.759                          -1.632            34.146238  
5                  18.436                          35.357            35.955973  
6                     NaN                             NaN            15.451376  
7                  18.027                          -9.970            15.473324  
8                  12.511                          12.575             7.322141

In [28]:
# Change NaN values to be represented by 0
BCG_df.fillna(0, inplace = True)

#### Total revenue and net income over the 3 years
##### The highest earning company over the three years is Apple, Microsoft and Tesla in 3rd place.

In [29]:
companies_summary = BCG_df.groupby("Company").agg({
    "Total Revenue ($M)": ["mean"],
    "Net Income ($M)": ["mean"],
}).round(3)

companies_summary

Total Revenue ($M) Net Income ($M)
                        mean            mean
Company                                     
Apple             389549.333       96844.667
Microsoft         218435.667       77745.000
Tesla              91975.000       11571.333

#### Net profit margin
##### A deep dive into the companies' net profit margin indicates that;
- Microsoft is the highest eaner with 36.6% in 2022 - 35.1% in 2024
- Followed by Apple with a range of 25.3% in 2022- 23.9% in 2024
- Teslas's was between 15.5%in 2022 - 7.3% in 2024

In [30]:
BCG_df['Net profit margin %'] = ((BCG_df['Net Income ($M)'] / BCG_df['Total Revenue ($M)']) *100).round(3)

BCG_df

Company  Year  Total Revenue ($M)  Net Income ($M)  Total Assets ($M)  \
0      Apple  2022              394328            99803             352755   
1      Apple  2023              383285            96995             352583   
2      Apple  2024              391035            93736             364980   
3  Microsoft  2022              198270            72738             364840   
4  Microsoft  2023              211915            72361             411976   
5  Microsoft  2024              245122            88136             512163   
6      Tesla  2022               81462            12587              82338   
7      Tesla  2023               96773            14974             106618   
8      Tesla  2024               97690             7153             122070   

   Total Liabilities ($M)  Cash Flow from Operating Activities ($M)  \
0                  302083                                    122151   
1                  290437                                    110543   
2                  308030                                    118254   
3                  198298                                     89035   
4                  205753                                     87582   
5                  243686                                    118548   
6                   36440                                     14724   
7                   43009                                     13256   
8                   48390                                     14923   

   Revenue_growth (%)  NetIncome_growth (%)  Assets_growth (%)  \
0               0.000                 0.000              0.000   
1              -2.800                -2.814             -0.049   
2               2.022                -3.360              3.516   
3               0.000                 0.000              0.000   
4               6.882                -0.518             12.920   
5              15.670                21.800             24.319   
6               0.000                 0.000              0.000   
7              18.795                18.964             29.488   
8               0.948               -52.231             14.493   

   Liabilities_growth (%)  Operations cashflow growth (%)  Net profit margin %  
0                   0.000                           0.000               25.310  
1                  -3.855                          -9.503               25.306  
2                   6.057                           6.976               23.971  
3                   0.000                           0.000               36.686  
4                   3.759                          -1.632               34.146  
5                  18.436                          35.357               35.956  
6                   0.000                           0.000               15.451  
7                  18.027                          -9.970               15.473  
8                  12.511                          12.575                7.322

In [31]:
# save changes in csv file
file_path = "BCG_Financials.csv"
BCG_df.to_csv(file_path, index = False)